src/mem/ruby/protocol/SConscript

```diff
@@ -34,87 +34,141 @@ from SCons.Scanner import Classic
 
 from gem5_scons import Transform
 
-Import('*')
+Import("*")
 
-if not env['CONF']['RUBY']:
+if not env["CONF"]["RUBY"]:
     Return()
 
-output_dir = Dir('.')
-html_dir = Dir('html')
-slicc_dir = Dir('../slicc')
+output_dir = Dir(".")
+html_dir = Dir("html")
+slicc_dir = Dir("../slicc")
 
-sys.path[1:1] = [ Dir('..').Dir('..').srcnode().abspath ]
+sys.path[1:1] = [Dir("..").Dir("..").srcnode().abspath]
 from slicc.parser import SLICC
 
 slicc_depends = []
-for root,dirs,files in os.walk(slicc_dir.srcnode().abspath):
+for root, dirs, files in os.walk(slicc_dir.srcnode().abspath):
     for f in files:
-        if f.endswith('.py'):
+        if f.endswith(".py"):
             slicc_depends.append(File(os.path.join(root, f)))
 
 #
 # Use SLICC
 #
 env["SLICC_PATH"] = env["PROTOCOL_DIRS"]
-slicc_scanner = Classic("SliccScanner", ['.sm', '.slicc'], "SLICC_PATH",
-                        r'''include[ \t]["'](.*)["'];''')
+slicc_scanner = Classic(
+    "SliccScanner",
+    [".sm", ".slicc"],
+    "SLICC_PATH",
+    r"""include[ \t]["'](.*)["'];""",
+)
 env.Append(SCANNERS=slicc_scanner)
 
-slicc_includes = ['mem/ruby/slicc_interface/RubySlicc_includes.hh'] + \
-        env['SLICC_INCLUDES']
-def slicc_emitter(target, source, env):
-    assert len(source) == 1
-    filepath = source[0].srcnode().abspath
-
-    slicc = SLICC(filepath, protocol_base.abspath, verbose=False)
-    slicc.process()
-    slicc.writeCodeFiles(output_dir.abspath, slicc_includes)
-    if env['CONF']['SLICC_HTML']:
-        slicc.writeHTMLFiles(html_dir.abspath)
-
-    target.extend([output_dir.File(f) for f in sorted(slicc.files())])
-    return target, source
-
-def slicc_action(target, source, env):
-    assert len(source) == 1
-    filepath = source[0].srcnode().abspath
+slicc_includes = ["mem/ruby/slicc_interface/RubySlicc_includes.hh"] + env[
+    "SLICC_INCLUDES"
+]
 
-    slicc = SLICC(filepath, protocol_base.abspath, verbose=True)
-    slicc.process()
-    slicc.writeCodeFiles(output_dir.abspath, slicc_includes)
-    if env['CONF']['SLICC_HTML']:
-        slicc.writeHTMLFiles(html_dir.abspath)
 
-slicc_builder = Builder(action=MakeAction(slicc_action, Transform("SLICC")),
-                        emitter=slicc_emitter)
-
-protocol = env['CONF']['PROTOCOL']
-protocol_dir = None
-for path in env['PROTOCOL_DIRS']:
-    if os.path.exists(path.File("%s.slicc" % protocol).abspath):
-        protocol_dir = Dir(path)
-        break
-
-if not protocol_dir:
-    raise ValueError("Could not find {}.slicc in PROTOCOL_DIRS".format(
-        protocol))
+def slicc_emitter(target, source, env):
```

Comment:
I didn't go through the next 80 lines *super* carefully since there is a lot I don't understand here, but if it has been tested with/without multiple protocols that's good enough for me to prove it works.

Reply:
Yeah, it's been tested, though it can always be tested more!

One thing to note here is that we have to run slicc on all protocols *twice*. The first time (the emitter code code here, IIRC) we just figure out what files we will be making (which, annoyingly requires an entire slicc parsing). Then, in the action code below we actually make the files. Unfortunately, this causes gem5's build to slow down significantly in the early stages (and it's not parallelizable at all).